### PR TITLE
Add logs function

### PR DESCRIPTION
**Pull request by campionfellin**
Examples:

`clasp logs`
```
(ERROR)    - myFunction      - oh no                - [Sat Apr 07 2018 10:58:31 GMT-0700 (PDT)]
(INFO)     - myFunction      - new                  - [Sat Apr 07 2018 10:58:31 GMT-0700 (PDT)]
(DEBUG)    - myFunction      - suppp                - [Sat Apr 07 2018 10:58:31 GMT-0700 (PDT)]
(ERROR)    - myFunction      - oh no                - [Sat Apr 07 2018 10:58:30 GMT-0700 (PDT)]
(INFO)     - myFunction      - new                  - [Sat Apr 07 2018 10:58:30 GMT-0700 (PDT)]
```

`clasp logs --json`
```
(ERROR)    - myFunction      - {
    "timestamp": {
        "seconds": 1523123911,
        "nanos": 894999980
    },
    "labels": {
        "script.googleapis.com/process_id": "maybe secret, not sure",
        "script.googleapis.com/user_key": "super secret key",
        "script.googleapis.com/project_key": "also very secret"
    },
    "insertId": "1yhq6jdg4coamx9",
    "httpRequest": null,
    "resource": {
        "labels": {
            "function_name": "myFunction",
            "project_id": "project-id-2758325992214167277",
            "invocation_type": "editor"
        },
        "type": "app_script_function"
    },
    "severity": "ERROR",
    "logName": "projects/project-id-2758325992214167277/logs/script.googleapis.com%2Fconsole_logs",
    "operation": null,
    "trace": "",
    "sourceLocation": null,
    "receiveTimestamp": {
        "seconds": "1523123912",
        "nanos": 901087233
    },
    "textPayload": "oh no",
    "payload": "textPayload"
} - [Sat Apr 07 2018 10:58:31 GMT-0700 (PDT)]
```

(Note: it also does the most recent 5 logs for json, but I've shortened it here)

`clasp logs --open`

Opens the logs in StackDriver in the browser.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -200,8 +204,13 @@ Forgot ${PROJECT_NAME} commands? Get help:\n  ${PROJECT_NAME} --help`,
   LOGGED_OUT: `\nCommand failed. Please login. (${PROJECT_NAME} login)`,
   OFFLINE: 'Error: Looks like you are offline.',
   ONE_DEPLOYMENT_CREATE: 'Currently just one deployment can be created at a time.',
+  NO_FUNCTION_NAME: 'N/A',
+  NO_GCLOUD_PROJECT: `\nPlease set your projectId in your .clasp.json file to your Google Cloud project ID. \n
+  You can find your projectId by following the instructions in the README here: \n
+  https://github.com/google/clasp/blob/master/README.md`,
```
Comment (grant) [inline index.ts]: Create an anchor link to the header that has the instructions

Like `https://github.com/google/clasp#projectid`.

File: index.ts [hunk 2]
```diff
@@ -200,8 +204,13 @@ Forgot ${PROJECT_NAME} commands? Get help:\n  ${PROJECT_NAME} --help`,
...
   NO_NESTED_PROJECTS: '\nNested clasp projects are not supported.',
   READ_ONLY_DELETE: 'Unable to delete read-only deployment.',
+  PAYLOAD_UNKNOWN: 'Unknown StackDriver payload',
```
Comment (grant) [inline index.ts]: Add period.

File: index.ts [hunk 3]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+          if (payloadType === 'textPayload') {
+            payload = entries[i].metadata.textPayload;
+          } else if (payloadType === 'jsonPayload') {
+            payload = entries[i].metadata.jsonPayload.fields.message.stringValue;
+          } else if (payloadType === 'protoPayload') {
+            payload = entries[i].metadata.protoPayload;
+          } else {
+            payload = ERROR.PAYLOAD_UNKNOWN;
+          }
+          if (payload && typeof(payload) === "string") {
```
Comment (grant) [inline index.ts]: Use `'` everywhere.

File: index.ts [hunk 4]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+        const logString = `${severity.padEnd(10)} - ${functionName} - ${payload} - [${timestamp}]`;
+        console.log(logString);       
+      }
+    }
+  
+    getProjectSettings().then(({ scriptId, rootDir, projectId }: ProjectSettings) => {
+      if (!projectId) {
+        console.error(ERROR.NO_GCLOUD_PROJECT);
+        process.exit(-1);
+      } else {
```
Comment (grant) [inline index.ts]: You can get rid of this `else` to reduce the nesting. Optionally `return` to explicitly tell the programmer we're stopping.

```ts
if (!projectId) {
  console.error(ERROR.NO_GCLOUD_PROJECT);
  return process.exit(-1);
}
if (cmd.open) {
```

File: index.ts [hunk 5]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+  .command('logs')
+  .description('Shows the StackDriver Logs')
+  .option('--json', "Show logs in JSON form")
+  .option('--open', 'Open the StackDriver logs in browser')
+  .action((cmd: LoginOptions) => {
+    function printLogs(entries) {
+      for (let i = 0; i < 5; i++) {
+        const severity = `(${entries[i].metadata.severity})`;
+        const timestamp = entries[i].metadata.timestamp;
+        const payloadType = entries[i].metadata.payload;
```
Comment (grant) [inline index.ts]: Destructure:
```ts
const { severtity, timestamp, payload } = metadata;
```

File: index.ts [hunk 6]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+        let functionName = entries[i].metadata.resource.labels.function_name;
+        if (functionName) {
+          functionName = functionName.padEnd(15);
+        } else {
+          functionName = ERROR.NO_FUNCTION_NAME;
+        }
+        let payload = '';
+        if (cmd.json) {
+          payload = JSON.stringify(entries[i], null, 4);
```
Comment (grant) [inline index.ts]: Use 2 instead of 4.

File: index.ts [hunk 7]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+      } else {
+        if (cmd.open) {
+          const stackdriverURL = `https://console.cloud.google.com/logs/viewer?project=${projectId}&resource=app_script_function`;
+          open(stackdriverURL);
+          process.exit(0);
+        }
+        const logging = new Logging({
+            projectId: projectId
+        });
+        logging.getEntries().then((entryData: any) => {
```
Comment (grant) [inline index.ts]: Modify printLogs to accept the data you get from getEntries. Then:

```ts
logging.getEntries().then(printLogs);
```

File: index.ts [hunk 8]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+ * Use --json for output in json format
+ * Use --open to open logs in StackDriver
+ */
+commander
+  .command('logs')
+  .description('Shows the StackDriver Logs')
+  .option('--json', "Show logs in JSON form")
+  .option('--open', 'Open the StackDriver logs in browser')
+  .action((cmd: LoginOptions) => {
+    function printLogs(entries) {
```
Comment (grant) [inline index.ts]: Get the first element ([array destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Array_destructuring)):
```ts
function printLogs([entries]) {
```

Requires usage of `printLogs` to be changed (see below).

File: index.ts [hunk 9]
```diff
@@ -1048,6 +1057,71 @@ commander
...
+          } else if (payloadType === 'protoPayload') {
+            payload = entries[i].metadata.protoPayload;
+          } else {
+            payload = ERROR.PAYLOAD_UNKNOWN;
+          }
+          if (payload && typeof(payload) === "string") {
+            payload = payload.padEnd(20);
+          }
+        }
+        const logString = `${severity.padEnd(10)} - ${functionName} - ${payload} - [${timestamp}]`;
```
Comment (grant) [inline index.ts]:
```ts
console.log(`${severity.padEnd(10)} - ${functionName} - ${payload} - [${timestamp}]`);
```

File: package.json [hunk 10]
```diff
@@ -37,6 +39,7 @@
     "find-parent-dir": "^0.3.0",
     "fs": "^0.0.1-security",
     "googleapis": "^25.0.0",
+    "grpc": "^1.10.1",
```
Comment (grant) [inline package.json]: Is grpc needed? Where do you use it?

File: index.ts [hunk 11]
```diff
@@ -1067,58 +1071,51 @@ commander
...
       }
-    });
+      if (cmd.open) {
+        const stackdriverURL = `https://console.cloud.google.com/logs/viewer?project=${projectId}&resource=app_script_function`;
+        open(stackdriverURL);
+        process.exit(0);
+      }
+      const logger = new logging({ projectId });
+      logger.getEntries().then(printLogs);
+      });
```
Comment (grant) [inline index.ts]: Nit: Indentation

File: package.json (hunk 10 above)
Comment (campionfellin) [reply package.json]: Oops, was having a weird error earlier and that fixed it. Seems to be fine without it now.

File: index.ts (hunk 11 above)
Comment (campionfellin) [reply index.ts]: Fixed.

Comment (grant) [reply index.ts]: This doesn't fix the indentation.
It should be like this:
```ts
  const logger = new logging({ projectId });
  logger.getEntries().then(printLogs);
});
```

(I'll have to make sure Travis catches this)

Comment (campionfellin) [reply index.ts]: Good catch, should be actually fixed now.

**Pull request merged**